keras_nlp/models/t5/t5_backbone.py

```diff
@@ -0,0 +1,846 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""T5 backbone model."""
+
+import copy
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.compiler.tf2xla.python.xla import dynamic_slice
+
+from keras_nlp.api_export import keras_nlp_export
+from keras_nlp.models.backbone import Backbone
+from keras_nlp.models.t5.t5_presets import backbone_presets
+from keras_nlp.utils.python_utils import classproperty
+
+
+@keras_nlp_export("keras_nlp.models.T5Backbone")
+class T5Backbone(Backbone):
+    def __init__(
+        self,
+        vocabulary_size,
+        output_dim,
+        hidden_dim,
+        key_value_dim,
+        num_blocks,
+        num_heads,
+        use_gated_activation,
+        activation,
+        dropout,
+        layer_norm_epsilon,
+        **kwargs,
+    ):
+        shared_embedding = keras.layers.Embedding(
+            input_dim=vocabulary_size,
+            output_dim=output_dim,
+            embeddings_initializer=keras.initializers.TruncatedNormal(1.0),
+            name="shared_embedding",
+        )
+        encoder = T5MainLayer(
```

Comment:
I think it would be a little more readable to consider removing "main layer" class entirely and switching to something like this...

```
encoder_embedding = token_embedding(encoder_token_ids)
decoder_embedding = token_embedding(decoder_token_ids)

# Set up masks.

for i in range(num_layers):
  # Call encoder blocks.

for in in range(num_layers):
  # Call decoder blocks.
```

There would be a few advantages: 
- There a lot of state passing that would be way easier to follow if there wasn't a shared encoder/decoder main class. E.g. all the "encoder state" `encoder_hidden_states`, `encoder_attention_mask`, `encoder_decoder_position_bias` and `encoder_layer_head_mask` that gets passed to the decoder.
- Model summary would look similar to other backbones.
- `get_layer` calls would work similar to other backbones.

The main thing the main layer seems to do is mask setup. We can just move that here and gradually use more utilities for it.

Reply:
Another way of saying this is that this implementation should look like our BART model unless there's a good reason to deviate.